File: packages/host4_flutter_aivoice/ios/AiVoice_副本/TokenGenerator/ByteBuf.m
//
//  ByteBuf.m
//  火山引擎 RTC Token Generator
//  字节缓冲区实现
//

#import "ByteBuf.h"

@interface ByteBuf ()

@property (nonatomic, strong) NSMutableData *buffer;
@property (nonatomic, assign) NSUInteger position;
@property (nonatomic, assign) NSUInteger readPosition;

@end

@implementation ByteBuf

- (instancetype)init {
    self = [super init];
    if (self) {
        _buffer = [NSMutableData dataWithCapacity:1024];
        _position = 0;
        _readPosition = 0;
    }
    return self;
}

- (instancetype)initWithBytes:(NSData *)bytes {
    self = [super init];
    if (self) {
        _buffer = [bytes mutableCopy];
        _position = bytes.length;
        _readPosition = 0;
    }
    return self;
}

- (NSData *)asBytes {
    return [NSData dataWithBytes:self.buffer.bytes length:self.position];
}

// MARK: - 写入方法

- (ByteBuf *)putShort:(int16_t)value {
    // Little Endian
    uint8_t bytes[2];
    bytes[0] = (uint8_t)(value & 0xFF);
    bytes[1] = (uint8_t)((value >> 8) & 0xFF);
    
    [self.buffer appendBytes:bytes length:2];
    self.position += 2;
    
    return self;
}

- (ByteBuf *)putInt:(int32_t)value {
    // Little Endian
    uint8_t bytes[4];
    bytes[0] = (uint8_t)(value & 0xFF);
    bytes[1] = (uint8_t)((value >> 8) & 0xFF);
    bytes[2] = (uint8_t)((value >> 16) & 0xFF);
    bytes[3] = (uint8_t)((value >> 24) & 0xFF);
    
    [self.buffer appendBytes:bytes length:4];
    self.position += 4;
    
    return self;
}

- (ByteBuf *)putLong:(int64_t)value {
    // Little Endian
    uint8_t bytes[8];
    bytes[0] = (uint8_t)(value & 0xFF);
    bytes[1] = (uint8_t)((value >> 8) & 0xFF);
    bytes[2] = (uint8_t)((value >> 16) & 0xFF);
    bytes[3] = (uint8_t)((value >> 24) & 0xFF);
    bytes[4] = (uint8_t)((value >> 32) & 0xFF);
    bytes[5] = (uint8_t)((value >> 40) & 0xFF);
    bytes[6] = (uint8_t)((value >> 48) & 0xFF);
    bytes[7] = (uint8_t)((value >> 56) & 0xFF);
    
    [self.buffer appendBytes:bytes length:8];
    self.position += 8;
    
    return self;
}

- (ByteBuf *)putBytes:(NSData *)value {
    // 先写入长度
    [self putShort:(int16_t)value.length];
    
    // 再写入数据
    [self.buffer appendData:value];
    self.position += value.length;
    
    return self;
}

- (ByteBuf *)putString:(NSString *)value {
    NSData *data = [value dataUsingEncoding:NSUTF8StringEncoding];
    return [self putBytes:data];
}

- (ByteBuf *)putIntMap:(NSDictionary<NSNumber *, NSNumber *> *)map {
    // 写入 map 大小
    [self putShort:(int16_t)map.count];
    
    // 按 key 排序（TreeMap 行为）
    NSArray *sortedKeys = [map.allKeys sortedArrayUsingSelector:@selector(compare:)];
    
    // 写入每个键值对
    for (NSNumber *key in sortedKeys) {
        NSNumber *value = map[key];
        [self putShort:[key shortValue]];
        [self putInt:[value intValue]];
    }
    
    return self;
}

// MARK: - 读取方法

- (int16_t)readShort {
    if (self.readPosition + 2 > self.buffer.length) {
        return 0;
    }
    
    const uint8_t *bytes = self.buffer.bytes;
    
    // Little Endian
    int16_t value = (int16_t)(bytes[self.readPosition] |
                             (bytes[self.readPosition + 1] << 8));
    
    self.readPosition += 2;
    
    return value;
}

- (int32_t)readInt {
    if (self.readPosition + 4 > self.buffer.length) {
        return 0;
    }
    
    const uint8_t *bytes = self.buffer.bytes;
    
    // Little Endian
    int32_t value = (int32_t)(bytes[self.readPosition] |
                             (bytes[self.readPosition + 1] << 8) |
                             (bytes[self.readPosition + 2] << 16) |
                             (bytes[self.readPosition + 3] << 24));
    
    self.readPosition += 4;
    
    return value;
}

- (NSData *)readBytes {
    int16_t length = [self readShort];
    
    if (length <= 0 || self.readPosition + length > self.buffer.length) {
        return [NSData data];
    }
    
    NSRange range = NSMakeRange(self.readPosition, length);
    NSData *data = [self.buffer subdataWithRange:range];
    
    self.readPosition += length;
    
    return data;
}

- (NSString *)readString {
    NSData *data = [self readBytes];
    return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
}

- (NSDictionary<NSNumber *, NSNumber *> *)readIntMap {
    int16_t length = [self readShort];
    
    NSMutableDictionary<NSNumber *, NSNumber *> *map = [NSMutableDictionary dictionary];
    
    for (int16_t i = 0; i < length; i++) {
        int16_t key = [self readShort];
        int32_t value = [self readInt];
        
        map[@(key)] = @(value);
    }
    
    return map;
}

@end

File: packages/host4_flutter_aivoice/ios/Classes/Utils/ByteBuf.swift
import Foundation

/// 字节缓冲区 — Little Endian 序列化/反序列化
class ByteBuf {
  private var buffer = Data(capacity: 1024)
  private var writePos = 0
  private var readPos = 0

  /// 空缓冲区初始化
  init() {}

  /// 从已有数据初始化（用于解析）
  init(data: Data) {
    buffer = data
    writePos = data.count
    readPos = 0
  }

  /// 获取已写入的数据
  var bytes: Data { buffer.prefix(writePos) }

  // MARK: - 写入

  @discardableResult
  func putShort(_ value: Int16) -> Self {
    var le = value.littleEndian
    withUnsafeBytes(of: &le) { buffer.append(contentsOf: $0) }
    writePos += 2
    return self
  }

  @discardableResult
  func putInt(_ value: Int32) -> Self {
    var le = value.littleEndian
    withUnsafeBytes(of: &le) { buffer.append(contentsOf: $0) }
    writePos += 4
    return self
  }

  @discardableResult
  func putLong(_ value: Int64) -> Self {
    var le = value.littleEndian
    withUnsafeBytes(of: &le) { buffer.append(contentsOf: $0) }
    writePos += 8
    return self
  }

  @discardableResult
  func putBytes(_ data: Data) -> Self {
    putShort(Int16(data.count))
    buffer.append(data)
    writePos += data.count
    return self
  }

  @discardableResult
  func putString(_ value: String) -> Self {
    putBytes(Data(value.utf8))
  }

  @discardableResult
  func putIntMap(_ map: [Int: Int]) -> Self {
    putShort(Int16(map.count))
    for key in map.keys.sorted() {
      putShort(Int16(key))
      putInt(Int32(map[key] ?? 0))
    }
    return self
  }

  // MARK: - 读取

  func readShort() -> Int16 {
    guard readPos + 2 <= buffer.count else { return 0 }
    let value = buffer.withUnsafeBytes { ptr in
      ptr.load(fromByteOffset: readPos, as: Int16.self).littleEndian
    }
    readPos += 2
    return value
  }

  func readInt() -> Int32 {
    guard readPos + 4 <= buffer.count else { return 0 }
    let value = buffer.withUnsafeBytes { ptr in
      ptr.load(fromByteOffset: readPos, as: Int32.self).littleEndian
    }
    readPos += 4
    return value
  }

  func readBytes() -> Data {
    let length = Int(readShort())
    guard length > 0, readPos + length <= buffer.count else { return Data() }
    let data = buffer.subdata(in: readPos..<readPos + length)
    readPos += length
    return data
  }

  func readString() -> String {
    String(data: readBytes(), encoding: .utf8) ?? ""
  }

  func readIntMap() -> [Int: Int] {
    let count = Int(readShort())
    var map: [Int: Int] = [:]
    for _ in 0..<count {
      let key = Int(readShort())
      let value = Int(readInt())
      map[key] = value
    }
    return map
  }
}
